namer/dnssrv/src/main/scala/io/buoyant/namer/dnssrv/DnsSrvNamer.scala

```diff
@@ -58,19 +66,26 @@ class DnsSrvNamer(
     }
   }
 
-  private def watchDns(dsnsrvRecord: String, timer: Timer): Activity[List[Address]] = {
-    val state = Var.async[Activity.State[List[Address]]](Activity.Pending) { update =>
+  private def watchDns(dnsSrvName: String, timer: Timer): Activity[List[Address]] = {
+    val iteration = new AtomicInteger(0)
+    val gauge = dnsIter.addGauge(dnsSrvName)(iteration.floatValue) // emit PER DNS iteration count, ref this val in the doUnit to avoid GC
 
+    val state: Var[Activity.State[List[Address]]] = Var.async[Activity.State[List[Address]]](Activity.Pending) { update =>
       def doUnit(): Unit = {
-        val lookup = new DNS.Lookup(dsnsrvRecord, DNS.Type.SRV, DNS.DClass.IN)
-        lookup.setResolver(resolver)
+        val lookup = new DNS.Lookup(dnsSrvName, DNS.Type.SRV, DNS.DClass.IN)
+        iteration.incrementAndGet()
+        lookup.setResolver(resolver())
+        lookup.setCache(null)
         Stat.time(latency)(lookup.run())
+        log.debug("thread %s id: %s dns: %s lookup %s iteration: %s gauge %s", Thread.currentThread().getName, Thread.currentThread().getId, dnsSrvName, lookup.toString, iteration, gauge)
         lookup.getResult match {
           case DNS.Lookup.HOST_NOT_FOUND | DNS.Lookup.TYPE_NOT_FOUND =>
-            val msg = s"no results for $dsnsrvRecord"
-            log.trace(msg)
+            val msg = s"no results for $dnsSrvName return type: ${lookup.getResult} errorString: ${lookup.getErrorString} iteration: ${iteration.intValue()}"
+            log.debug(s"in doUnit: message %s", msg)
             failure.incr()
+            dnsFailed.scope(lookup.getErrorString).scope(dnsSrvName).counter("total").incr()
```

Comment:
Caution that the number of counters could blow up here if the number of distinct error strings is large.  if there are only a few potential errors messages, it's probably fine.

Also note that looking up/creating counters like this can be costly, especially in the hot path.  It may not be an issue, but if performance is a concern here, consider caching the counters so you don't have to keep creating them.

Reply:
Even if the number is bounded, this is potentially a lot of metrics.  How important is it to scope these counters by dns entry?

Reply:
Sorry for the late response - I do agree the DNS entries is not 100% necessary and it can cause problem - I will remove them